GVFS/GVFS.Common/InstallerPreRunChecker.cs

```diff
@@ -0,0 +1,212 @@
+using GVFS.Common;
+using GVFS.Common.Git;
+using GVFS.Common.Tracing;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Threading;
+
+namespace GVFS.Upgrader
+{
+    public class InstallerPreRunChecker
+    {
+        private static readonly HashSet<string> BlockingProcessSet = new HashSet<string> { "GVFS", "GVFS.Mount", "git", "ssh-agent", "bash", "wish", "git-bash" };
+
+        private ITracer tracer;
+                
+        public InstallerPreRunChecker(ITracer tracer)
+        {
+            this.tracer = tracer;
+            this.CommandToRerun = string.Empty;
+        }
+
+        public string CommandToRerun { get; set; }
+
+        public bool TryRunPreUpgradeChecks(out string consoleError)
+        {
+            this.tracer.RelatedInfo("Checking if GVFS upgrade can be run on this machine.");
+
+            if (this.IsUnattended())
+            {
+                consoleError = "`gvfs upgrade` is not supported in unattended mode";
+                this.tracer.RelatedError($"{nameof(TryRunPreUpgradeChecks)}: {consoleError}");
+                return false;
+            }
+
+            if (this.IsDevelopmentVersion())
+            {
+                consoleError = "Cannot run upgrade when development version of GVFS is installed.";
+                this.tracer.RelatedError($"{nameof(TryRunPreUpgradeChecks)}: {consoleError}");
+                return false;
+            }
+
+            if (!this.IsGVFSUpgradeAllowed(out consoleError))
+            {
+                this.tracer.RelatedError($"{nameof(TryRunPreUpgradeChecks)}: {consoleError}");
+                return false;
+            }
+
+            this.tracer.RelatedInfo("Successfully finished pre upgrade checks. Okay to run GVFS upgrade.");
+
+            consoleError = null;
+            return true;
+        }
+        
+        // TODO: Move repo mount calls to GVFS.Upgrader project.
+        public bool TryMountAllGVFSRepos(out string consoleError)
+        {
+            return this.TryRunGVFSWithArgs("service --mount-all", out consoleError);
+        }
+
+        public bool TryUnmountAllGVFSRepos(out string consoleError)
+        {
+            consoleError = null;
+
+            this.tracer.RelatedInfo("Unmounting any mounted GVFS repositories.");

```

Comment:
Instead of this pattern of tracing "Starting foo" and and the end "Finished foo", call `ITracer.StartActivity` and put it in a using statement. It'll automatically log the start and stop messages, but also record the duration of the activity. Please apply it to any place you have this pattern.